### ISSUE TITLE
twisted.web.static.loadMimeTypes overrides mime.types default location for no good reason

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5717]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5717|
|Type|defect|
|Created|2012-06-19 19:30:00Z|
|Branch|https://github.com/twisted/twisted/tree/loadMimeTypes-5717|

The Python `mimetypes` module has an `init` function which will read `mime.types` files from a bunch of likely locations throughout the system.  Twisted Web hard-codes the default to be just `/etc/mime.types`.  We don't know better than the stdlib in this case, and two hard-coded strings is worse than one, so we should just call `mimetypes.init()` and forget about it.

<details><summary>Searchable metadata</summary>

```
trac-id__5717 5717
type__defect defect
reporter__glyph glyph
priority__low low
milestone__ 
branch__branches_loadMimeTypes_5717 branches/loadMimeTypes-5717
branch_author__lvh lvh
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1340134200000000 1340134200000000
changetime__1378639561000000 1378639561000000
version__None None
owner__lvh lvh
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_2"></a>|@lvh commented|
|-|-|

(In [39287]) Branching to 'loadMimeTypes-5717'

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_3"></a>|@lvh commented|
|-|-|

I believe to have fixed this in the branch in the way glyph suggests.

By virtue of using ```mimetypes.init``` (in this branch), this fixes [#5170](https://github.com/twisted/twisted/issues/5170) for free. I have marked that as a duplicate.

Buildbots: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/loadMimeTypes-5717

The mimetype_locations argument is unchanged, but does not follow the coding standard. Can I rename it?

As a bonus: this actually fixes my problem, where an ```<img src="something.svg">``` tag didn't actually display the SVG, because it was being served as ```text/plain```. Turns out my OS is smart enough to know what SVG is. Woo!

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @radix<br>@radix set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_5"></a>|@radix set owner to @lvh<br>@radix set status to `new`|
|-|-|

Looks good. I would advise against changing mimetypes_locations, since it is a keyword argument and very likely is already being passed as one.

The test for _init is a bit bizarre, but I can see the appeal given that it doesn't require mocking global state.

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Thanks for the review, radix!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

For the sake of consistency, I have told a lot of other people that `_`-prefixed positional arguments are not a viable strategy for making private arguments.  Nothing prevents users from writing code like:

```
    loadMimeTypes(foo, bar)
```

And nothing here indicates to the user that they shouldn't be doing this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

To be clear, since the parameter isn't actually private, I suggest not naming it in a way that confuses people (maintainers included) so they they may mistakenly believe that it is private.

Whether further changes are necessary, I don't know.  That depends how comfortable you are introducing this new public interface.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_9"></a>|@lvh commented|
|-|-|

Why isn't it private? It's private and testing-only, both in practice and in documentation.

I'd normally use a class attribute instead, but this thing isn't a class. What do you suggest?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Look at this:

```
    loadMimeTypes(foo, bar)
```

Where are the leading underscores?  Why would anything here be private?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Replying to [lvh](#note_9):

> I'd normally use a class attribute instead, but this thing isn't a class. What do you suggest?

I'd suggest just making it public.  There are many reasons to want to make _attributes_ private by default:

  - additional attributes add additional burden to other implementors trying to provide a substitutable object
  - additional attributes imply a commitment to a particular implementation strategy that relies upon those attributes
  - additional public attributes provide access to things that may be internal details that we'd like to preserve the flexibility to change

And even so, I still think there are certain attributes (like 'reactor') that we really need to start thinking about making public by default because it's too hard for applications outside Twisted to test certain objects without fiddling with their private state.

However, there's _really_ no point to having a function make its _parameters_ private.  If someone wants to pass a decorated version of mimetypes.init to this function, what's the benefit of even trying to stop them?  It's probably not too useful, the documentation shouldn't emphasize it, but if someone wants to do it there's no particular maintenance burden to keeping that parameter around forever (we don't necessarily have to _use_ it; the behavior could deprecate passing it, and change it in the future to be ignored, but code passing it wouldn't start raising an exception, it would just do something different).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_12"></a>|@lvh commented|
|-|-|

Okay, I've made it public.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj set owner to @lvh<br>@wallrj set status to `new`|
|-|-|

Thanks lvh,

Notes:
 * Forced build - all pass
 * Merges cleanly
 * On fedora 19 I get 1081 mimetypes as opposed to 120 which is great!
 * Although I note that it still doesn't include .mkv which someone
   was asking me about in #twisted-web yesterday.
   * I tested that I can add .mkv that to a custom mime-types
     file and pass it to loadMimeTypes. And it gets appended to the
     default types as expected.

Points:
 1. [branches/loadMimeTypes-5717/twisted/web/static.py](branches/loadMimeTypes-5717/twisted/web/static.py)
    1. You've left the hardcoded list of types in there.
    2. Some of those types seem to be already specified or specified
       differently in mimetypes
       ```
       In [1]: import mimetypes

       In [2]: mimetypes.init()

       In [3]: twistedTypes =         {
          ...:             '.conf':  'text/plain',
          ...:             '.diff':  'text/plain',
          ...:             '.exe':   'application/x-executable',
          ...:             '.flac':  'audio/x-flac',
          ...:             '.java':  'text/plain',
          ...:             '.ogg':   'application/ogg',
          ...:             '.oz':    'text/x-oz',
          ...:             '.swf':   'application/x-shockwave-flash',
          ...:             '.tgz':   'application/x-gtar',
          ...:             '.wml':   'text/vnd.wap.wml',
          ...:             '.xul':   'application/vnd.mozilla.xul+xml',
          ...:             '.py':    'text/plain',
          ...:             '.patch': 'text/plain',
          ...:         }

        In [6]: [(k, v, mimetypes.types_map.get(k)) for k, v in twistedTypes.items()]
        Out[6]:
        [('.xul',
          'application/vnd.mozilla.xul+xml',
          'application/vnd.mozilla.xul+xml'),
         ('.patch', 'text/plain', None),
         ('.diff', 'text/plain', None),
         ('.py', 'text/plain', 'text/x-python'),
         ('.conf', 'text/plain', 'text/plain'),
         ('.oz', 'text/x-oz', None),
         ('.java', 'text/plain', None),
         ('.wml', 'text/vnd.wap.wml', 'text/vnd.wap.wml'),
         ('.tgz', 'application/x-gtar', 'application/gzip'),
         ('.flac', 'audio/x-flac', 'audio/x-flac'),
         ('.swf', 'application/x-shockwave-flash', 'application/x-shockwave-flash'),
         ('.ogg', 'application/ogg', 'audio/ogg'),
         ('.exe', 'application/x-executable', 'application/octet-stream')]
        ```

    3. I suggest removing the ones that are already specified.

    4. Also I note that the function has always updated the global
       mimetypes.types_map dictionary. I suggest returning a
       copy of that dictionary. eg

       contentTypes = dict(...hardcoded twisted defaults...)
       contentTypes.update(mimetypes.types_map)
       return contentTypes

 2. [branches/loadMimeTypes-5717/twisted/web/test/test_static.py](branches/loadMimeTypes-5717/twisted/web/test/test_static.py)
    1. test_extraLocationsWork
       1. Maybe it would be clearer to use a simple object() sentinel
          here too, because using a list initially made me question
          why you were using assertIdentical instead of assertEqual
          here. Doesn't really matter I suppose.

    2. test_usesGlobalInitFunction
       1. Its clever how you get the default argument value.
       2. But isn't this a good example of where you could use the
          class based testing pattern described by Itamar here:
          http://blog.futurefoundries.com/2013/04/unittesting-without-patching.html
       3. That would avoid having to add a new public argument to
          loadMimeTypes too.

Radix already gave a +1, so I'm happy for you to merge this if you
disagree with the points above.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_16"></a>|@lvh commented|
|-|-|

Replying to [rwall](#note_15):
> Thanks lvh,

Thank you!

>    * I tested that I can add .mkv that to a custom mime-types
>      file and pass it to loadMimeTypes. And it gets appended to the
>      default types as expected.

Would you like me to add MKV?

> Points:
>  1. [branches/loadMimeTypes-5717/twisted/web/static.py](branches/loadMimeTypes-5717/twisted/web/static.py)
>     1. You've left the hardcoded list of types in there.

Yes. Should it be removed?

>     2. Some of those types seem to be already specified or specified
>        differently in mimetypes
>        ```
>        In [1]: import mimetypes
> 
>        In [2]: mimetypes.init()
> 
>        In [3]: twistedTypes =         {
>           ...:             '.conf':  'text/plain',
>           ...:             '.diff':  'text/plain',
>           ...:             '.exe':   'application/x-executable',
>           ...:             '.flac':  'audio/x-flac',
>           ...:             '.java':  'text/plain',
>           ...:             '.ogg':   'application/ogg',
>           ...:             '.oz':    'text/x-oz',
>           ...:             '.swf':   'application/x-shockwave-flash',
>           ...:             '.tgz':   'application/x-gtar',
>           ...:             '.wml':   'text/vnd.wap.wml',
>           ...:             '.xul':   'application/vnd.mozilla.xul+xml',
>           ...:             '.py':    'text/plain',
>           ...:             '.patch': 'text/plain',
>           ...:         }
> 
>         In [6]: [(k, v, mimetypes.types_map.get(k)) for k, v in twistedTypes.items()]
>         Out[6]:
>         [('.xul',
>           'application/vnd.mozilla.xul+xml',
>           'application/vnd.mozilla.xul+xml'),
>          ('.patch', 'text/plain', None),
>          ('.diff', 'text/plain', None),
>          ('.py', 'text/plain', 'text/x-python'),
>          ('.conf', 'text/plain', 'text/plain'),
>          ('.oz', 'text/x-oz', None),
>          ('.java', 'text/plain', None),
>          ('.wml', 'text/vnd.wap.wml', 'text/vnd.wap.wml'),
>          ('.tgz', 'application/x-gtar', 'application/gzip'),
>          ('.flac', 'audio/x-flac', 'audio/x-flac'),
>          ('.swf', 'application/x-shockwave-flash', 'application/x-shockwave-flash'),
>          ('.ogg', 'application/ogg', 'audio/ogg'),
>          ('.exe', 'application/x-executable', 'application/octet-stream')]
>         ```
> 
>     3. I suggest removing the ones that are already specified.

Isn't this list dependent on your system's opinion on MIME types?

>     4. Also I note that the function has always updated the global
>        mimetypes.types_map dictionary. I suggest returning a
>        copy of that dictionary. eg
> 
>        contentTypes = dict(...hardcoded twisted defaults...)
>        contentTypes.update(mimetypes.types_map)
>        return contentTypes

That's a good point and certainly looks like an improvement.

>  2. [branches/loadMimeTypes-5717/twisted/web/test/test_static.py](branches/loadMimeTypes-5717/twisted/web/test/test_static.py)
>     1. test_extraLocationsWork
>        1. Maybe it would be clearer to use a simple object() sentinel
>           here too, because using a list initially made me question
>           why you were using assertIdentical instead of assertEqual
>           here. Doesn't really matter I suppose.

Maybe. If I would change it, assertEqual seems more correct...

>     2. test_usesGlobalInitFunction
>        1. Its clever how you get the default argument value.
>        2. But isn't this a good example of where you could use the
>           class based testing pattern described by Itamar here:
>           http://blog.futurefoundries.com/2013/04/unittesting-without-patching.html
>        3. That would avoid having to add a new public argument to
>           loadMimeTypes too.

Hadn't seen that blog post before, thanks! It seems like it's more talking about the evils of patching vs injection, but what I'm doing is already injection, right? I suppose it would have the benefit that I need less hacks to get the default value out, and it would fix the second argument. I think the latter is slightly more compelling than the former, and I'm on the fence if the added complexity of the implementation is worth it. Assuming I'm not misunderstanding about the patching vs injection part, and you feel it's better, I'd be happy to write it.

> Radix already gave a +1, so I'm happy for you to merge this if you
> disagree with the points above.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj commented|
|-|-|

Replying to [lvh](#note_16):
> Would you like me to add MKV?

No, I don't think we should add any more fixed types to that list. I
used that extension as an example to test that users can still
override the types if they want to.

> > Points:
> >  1. [branches/loadMimeTypes-5717/twisted/web/static.py](branches/loadMimeTypes-5717/twisted/web/static.py)
> >     1. You've left the hardcoded list of types in there.
> Yes. Should it be removed?

I think it's a shame it was added in the first place. But it would be
backwards incompatible to remove it now, right?

> >         In [6]: [(k, v, mimetypes.types_map.get(k)) for k, v in twistedTypes.items()]
> >         Out[6]:

> >          ('.py', 'text/plain', 'text/x-python'),
 * https://tools.ietf.org/html/rfc2046#section-4.1.4

> >          ('.tgz', 'application/x-gtar', 'application/gzip'),
* http://www.ietf.org/rfc/rfc6713.txt

> >          ('.ogg', 'application/ogg', 'audio/ogg'),
 * https://www.rfc-editor.org/rfc/rfc5334.txt

> >          ('.exe', 'application/x-executable', 'application/octet-stream')]
 * https://tools.ietf.org/html/rfc2046#section-4.5.3

> >     3. I suggest removing the ones that are already specified.
> Isn't this list dependent on your system's opinion on MIME types?

I'd remove just the items which differ from the mime-types module and
which are standardised (the ones highlighted above)

See also:
 * https://www.iana.org/assignments/media-types

> ...I'm on the fence if the added
> complexity of the implementation is worth it. Assuming I'm not
> misunderstanding about the patching vs injection part, and you feel
> it's better, I'd be happy to write it.

On reflection I agree. It would add unnecessary complexity.

> > Radix already gave a +1, so I'm happy for you to merge this if you
> > disagree with the points above.

Please merge after removing the standardised mime-types.

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_18"></a>|@lvh set status to `closed`|
|-|-|

(In [39880]) Merge loadMimeTypes-5717: use available system MIME types

Author: lvh
Reviewer: rwall
Fixes: [#5717](https://github.com/twisted/twisted/issues/5717)

MIME type definition file path used to be hardcoded. Now, we rely on the stdlib to discover MIME types the system knows about. This greatly increases the number of known MIME types on most platforms.
